DQM: Restructure QTests

schneiml opened the pull request:
#### PR description:

This PR aims to clean up the way QTests are handled in DQM. Rather than spreading the logic all over the place, it is now all in a single module.
- `QualityTester` becomes a plugin in `DQMServices/Components`, handling all QTest things.
- The `DQMServices/ClientConfig` package disappears without replacement.
- Some useful parts of `ClientConfig` are rescued over to `Components`, most prominently `DQMGenericClient`.
- Some modules that used to use the XML parsing helpers from `ClientConfig` are now stubbed out (`SiPixelMonitorClient`/`SiStripMonitorClient`/`SiStripCommissioningClients`). Subsystem contacted to see if they need to be/can be ported.
- The XML parsing is now done with `boost::property_tree` rather than `xerces`. Much less code, however more strict (some of the QTest configuration XML used in production was malformed!) and probably worse error reporting.
- The QTest implementation remain in `DQMServices/Core` `QTest.[h|cc]`. However, the *only* place that uses them now is `QualityTester`.
- `QReport`s (results) remain available via the `MonitorElement`, but don't contain pointers to `QCriterion`s (the actual tests) any more.
- `QualityTester` is now a `DQMEDHarvester`, rather than a legacy module.
- The ordering `QualityTester` vs. other harvesters is not enforced by dependencies. This may need to be fixed.

#### PR validation:

Runs a simple workflow. Opening this PR for more validation, while some functionality is still missing (XML parsing for tracker clients).

Need detailed comparisons over many workflows to see if all QTests are still applied in the same way There might be XML parsing differences, and the pattern matching is different in the new code (now using POSIX `fnmatch` rather than `classlib`).

schneiml committed:
Move DQMGenericClient and QualityTester

... to DQMServices/Components/plugins, with the goal of eliminating the
ClientConfig package entirely.

Also merge the headers into .cc and remove the ROOT5 support code in
DQMGenericClient.

schneiml committed:
Remove HLTTauRelvalQTester.

It looks rather obscure and unused and is the only user of
QualityTester-as-a-library.

schneiml committed:
Make DQMRivetClient independent of DQMGenericClient

It was the only used of DQMGenericClient-as-a-library, for the
NormOptions struct, which is no duplicated there.

schneiml committed:
Merge FitSlicesYTool into DQMGenericClient.

As far as I see the remaining usages are forgotten imports.

schneiml committed:
Remove spurious usages of FitSlicesYTool.

schneiml committed:
Remove the ClientConfig package.

Some usages remain, but most references where probably unused anyways.

schneiml committed:
FIXUP: SealModule for QualityTester

schneiml committed:
Replace QualityTester implementation with stubs

Compiles now but does not do anything. Also moved it from legacy to
DQMEDHarvester, which will cause fallout over a lot of packages.

schneiml committed:
Stub out SiStripCommissioningClients SummaryPlotXmlParser

schneiml committed:
Stub out xml stuff in SiPixelMonitorClient.

schneiml committed:
Stub out xml stuff in SiStripMonitorClient.

schneiml committed:
Inline efficSet at its single usage site.

schneiml committed:
Provide an config alias for QualityTester

schneiml committed:
Automatically convert EDAnalyzer("QualityTester")

... to new alias.
Command was
sed -E -i 's/^(.*)cms.EDAnalyzer..QualityTester..(.*)$/from DQMServices.Core.DQMQualityTester import DQMQualityTester\n\1DQMQualityTester(\2/g' $(ag -l QualityTester --py)

schneiml committed:
Add missing module declaration to DQMGenericClient

schneiml committed:
New QTest XML file parsing.

Not complete yet, and not tested at all, but it compiles, at least.

schneiml committed:
Code-format.

schneiml committed:
FIx malformed XML for Qtests.

Some QTest files simply use the wrong tag name, and that inconsistently
between opening and closing tags. Surprising that the old parser
accepted that...

schneiml committed:
Change datastructure in QualityTester.

schneiml committed:
Remove a bunch of QTest code in DQMStore etc.

DQMStore no longer handles QTests. QReport no longer points to the
QCriterion. This invloves a bunch of API changes, but none of the APIs
where actually used.

Now QualityTester should be the only module interacting with QCriterion
and QTest.h.

schneiml committed:
Clean up BuildFiles.

schneiml committed:
Add remaining parameters and types.

The parameter names are terribly inconsistent, I extracted the mappings
form the old ClientConfig/src/QTestConfigure.cc. I set very few
defaults, lte's see how many errors that causes.

schneiml committed:
Implement actually running QTests.

Seems to work fine, for a single workflow at least.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12557

- This PR adds an extra 380KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQMServices/ClientConfig/interface/DQMGenericClient.h  modified in PR(s): #28247
  - File  DQMServices/Components/plugins/SealModule.cc  modified in PR(s): #28295
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #28092, #28247
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #28092, #28247
  - File  DQMServices/Core/interface/QReport.h  modified in PR(s): #28092
  - File  DQMServices/Core/interface/QTest.h  modified in PR(s): #28092
  - File  DQMServices/Core/src/DQMStore.cc  modified in PR(s): #28092, #28247
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #28092, #28247

cmsbuild commented:
A new Pull Request was created by  @schneiml (Marcel Schneider) for master.

It involves the following packages:

CalibMuon/DTCalibration
DQM/BeamMonitor
DQM/CSCMonitorModule
DQM/DTMonitorClient
DQM/DTMonitorModule
DQM/HLTEvF
DQM/HLXMonitor
DQM/HcalTasks
DQM/Integration
DQM/L1TMonitor
DQM/L1TMonitorClient
DQM/RPCMonitorClient
DQM/SiPixelCommon
DQM/SiPixelMonitorClient
DQM/SiPixelPhase1Config
DQM/SiPixelPhase1Summary
DQM/SiStripCommissioningClients
DQM/SiStripMonitorClient
DQM/SiStripMonitorSummary
DQM/TrackingMonitorClient
DQMOffline/CalibMuon
DQMOffline/Ecal
DQMOffline/JetMET
DQMOffline/L1Trigger
DQMOffline/Muon
DQMOffline/Trigger
DQMServices/ClientConfig
DQMServices/Components
DQMServices/Core
DQMServices/Examples
GeneratorInterface/RivetInterface
HLTriggerOffline/Common
HLTriggerOffline/Exotica
HLTriggerOffline/Higgs
HLTriggerOffline/JetMET
HLTriggerOffline/Muon
HLTriggerOffline/SUSYBSM
HLTriggerOffline/Tau
PhysicsTools/NanoAOD
Validation/L1T
Validation/RPCRecHits
Validation/RecoMuon
Validation/RecoTrack


@SiewYan, @andrius-k, @schneiml, @Martin-Grunewald, @rekovic, @fioriNTU, @tlampen, @alberto-sanchez, @pohsun, @santocch, @peruzzim, @cmsbuild, @agrohsje, @fwyzard, @efeyazgan, @tocheng, @jfernan2, @qliphy, @benkrikler, @mkirsano, @kmaeshima, @christopheralanwest, @franzoni, @fgolf can you please review it and eventually sign? Thanks.
@erikbutz, @rappoccio, @felicepantaleo, @jandrea, @missirol, @argiro, @fioriNTU, @alberto-sanchez, @ahinzmann, @threus, @mmusich, @seemasharmafnal, @venturia, @acimmino, @mmarionncern, @kreczko, @hdelanno, @battibass, @makortel, @jhgoh, @jdolen, @HuguesBrun, @agrohsje, @cericeci, @ptcox, @trocino, @rociovilar, @abbiendi, @barvic, @DryRun, @GiacomoSguazzoni, @tocheng, @VinInn, @jdamgov, @bellan, @nhanvtran, @gkasieczka, @rovere, @schoef, @idebruyn, @ebrondol, @mtosi, @dgulhan, @clelange, @mkirsano, @batinkov, @Fedespring, @calderona, @wmtford, @gpetruc, @mariadalfonso, @folguera this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

schneiml commented:
please test

... to see what happens.

For actual integration, I should probably factor out all the configuration changes to get the number of changed files down a bit.

cmsbuild commented:
-1

Tested at: 2b562a9f051218b319b6995380e67fecb9b68ba8

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f137dc/3271/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
9.0 step4
<pre>runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step4_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log

</pre>25.0 step4
<pre>runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/step4_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT.log

</pre>4.53 step4
<pre>runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT/step4_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT.log

</pre>136.731 step4
<pre>runTheMatrix-results/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/step4_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2.log

</pre>1000.0 step4
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step4_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

schneiml committed:
Fix more malformed XML

cmsbuild commented:
The code-checks are being triggered in jenkins.

schneiml commented:
All the failures where (for now) caused by a missing closing tag in `DQM/L1TMonitorClient/data/L1EmulatorCsctfQualityTests.xml`.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12572

- This PR adds an extra 384KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQMServices/ClientConfig/interface/DQMGenericClient.h  modified in PR(s): #28247
  - File  DQMServices/Components/plugins/SealModule.cc  modified in PR(s): #28295
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #28247, #28092
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #28247, #28092
  - File  DQMServices/Core/interface/QReport.h  modified in PR(s): #28092
  - File  DQMServices/Core/interface/QTest.h  modified in PR(s): #28092
  - File  DQMServices/Core/src/DQMStore.cc  modified in PR(s): #28247, #28092
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #28247, #28092

cmsbuild commented:
Pull request #28316 was updated. @SiewYan, @andrius-k, @schneiml, @Martin-Grunewald, @rekovic, @fioriNTU, @tlampen, @alberto-sanchez, @pohsun, @santocch, @peruzzim, @cmsbuild, @agrohsje, @fwyzard, @efeyazgan, @tocheng, @jfernan2, @qliphy, @benkrikler, @mkirsano, @kmaeshima, @christopheralanwest, @franzoni, @fgolf can you please check and sign again.

schneiml commented:
please test

maybe now we can get all the way to a comparison.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: de9e73c42a8f536af0d3e2ae731ad314917842d0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f137dc/3283/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f137dc/3283/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2932096
 - DQMHistoTests: Total failures: 152
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2931603
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -8145.719 KiB( 33 files compared)
 - DQMHistoSizes: changed ( 1000.0,... ): -81.534 KiB      RPC/AllHits
 - DQMHistoSizes: changed ( 1000.0,... ): -78.862 KiB      RPC/Muon
 - DQMHistoSizes: changed ( 1000.0,... ): -71.587 KiB      Pixel/Barrel
 - DQMHistoSizes: changed ( 1000.0,... ): -43.619 KiB      Pixel/Endcap
 - DQMHistoSizes: changed ( 1000.0 ): -18.192 KiB      SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 1000.0,... ): -16.682 KiB      Pixel/AdditionalPixelErrors
 - DQMHistoSizes: changed ( 1000.0,... ): -4.930 KiB       SiStrip/IsolatedBunches
 - DQMHistoSizes: changed ( 1000.0,... ): -3.624 KiB       L1TEMU/L1TdeRCT_FromRCT
 - DQMHistoSizes: changed ( 1000.0,... ): -3.350 KiB       Tracking/TrackParameters
 - DQMHistoSizes: changed ( 1000.0,... ): -3.296 KiB       L1TEMU/L1TdeRCT
 - DQMHistoSizes: changed ( 1000.0 ):  ... <truncated>
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

schneiml commented:
Looks like a bunch of QTests went missing with these changes -- not too surprising, I'll look at the details tomorrow. At least, the summaries seem to *mostly* still work, apart from some changes due to the missing (or maybe changed?) QTests.

schneiml committed:
Import DQMEDHarvester from new DQMStore branch.

Compared to the old code, this properly uses tokens, and exposes a few
more callbacks to allow porting more legacy modules.

schneiml commented:
The problem here is that most configs set the QTests to run on `endRun` (*only*), but the `DQMEDHarvester` does not run on `endRun`. So only those few that also enable endJob get produced.

I am looking for a good solution; running all Qtests in `endJob` is not good since then things depend on order of execution of the modules.

schneiml committed:
Enable QTest on endRun.

Also use the correct option for fnmatch, and turn on end-job tests by
default.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12594

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQMServices/ClientConfig/interface/DQMGenericClient.h  modified in PR(s): #28247
  - File  DQMServices/Components/plugins/SealModule.cc  modified in PR(s): #28295
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #28092, #28247
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #28092, #28247
  - File  DQMServices/Core/interface/QReport.h  modified in PR(s): #28092
  - File  DQMServices/Core/interface/QTest.h  modified in PR(s): #28092
  - File  DQMServices/Core/src/DQMStore.cc  modified in PR(s): #28092, #28247
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #28092, #28247

cmsbuild commented:
Pull request #28316 was updated. @SiewYan, @andrius-k, @schneiml, @Martin-Grunewald, @rekovic, @fioriNTU, @tlampen, @alberto-sanchez, @pohsun, @santocch, @peruzzim, @cmsbuild, @agrohsje, @fwyzard, @efeyazgan, @tocheng, @jfernan2, @qliphy, @benkrikler, @mkirsano, @kmaeshima, @christopheralanwest, @franzoni, @fgolf can you please check and sign again.

schneiml commented:
Now things should be better. I updated the `DQMEDHarvester` base class using the respective one form the "new DQMStore" branch, which means that also harvesting in general could be broken, but we get some better use of dependencies and more callbacks, including `dqmEndRun` (actually, I just added that one...).

The other big source of missing QTests was that we indeed want `*` to match `/` as well.

Overall, order of execution is a big mess with qtests. By running them in `endRun` we can be sure they are done by the time harvesting reads them in `endJob`, but then we might want to have Qtests on MEs booked in `endJob`, so we need to run the Qtests again. Actually, `MuonRecoTest` [1] books MEs and then expects the QTests on them to be populated immediatly. Not sure if that ever worked, it sure won't now.

In `endRun` we could enforce order by consuming and producing Tokens, but for `endJob`, we are doomed to have order encoded in sequences, afaik.

[1] https://cmssdt.cern.ch/dxr/CMSSW/source/DQMOffline/Muon/src/MuonRecoTest.cc#57

schneiml commented:
please test

... hope it all builds fine, the Harvesting code causes a number of compiler warnings...

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 044d8771786aa21f017760bd69acbaa2a432467d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f137dc/3305/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f137dc/3305/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2932096
 - DQMHistoTests: Total failures: 48
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2931707
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -101.899 KiB( 33 files compared)
 - DQMHistoSizes: changed ( 1000.0,... ): -71.587 KiB      Pixel/Barrel
 - DQMHistoSizes: changed ( 1000.0,... ): -43.619 KiB      Pixel/Endcap
 - DQMHistoSizes: changed ( 1000.0,... ): -16.682 KiB      Pixel/AdditionalPixelErrors
 - DQMHistoSizes: changed ( 1000.0,... ): 0.533 KiB        Tracking/TrackParameters
 - DQMHistoSizes: changed ( 1000.0,... ): -0.093 KiB       L1T/L1TOccupancy
 - DQMHistoSizes: changed ( 10024.0,... ): -14.303 KiB      HLT/Higgs
 - DQMHistoSizes: changed ( 11634.0,... ): -35.896 KiB      HLT/Higgs
 - DQMHistoSizes: changed ( 136.731 ): 0.063 KiB        SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 136.788 ): 0.250 KiB        SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 136.85 ): 1.112 KiB        SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 140.56 ):  ... <truncated>
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

schneiml committed:
Implement SiPixel config reading.

cmsbuild commented:
The code-checks are being triggered in jenkins.

schneiml commented:
Nothing to see here, except some new, untested code for `SiPixelActionExecutor` (which is responsible for many of the missing plots). Somehow screwed up my dev environment, saving changes here...

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12614

- This PR adds an extra 488KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQM/TrackingMonitorClient/plugins/TrackingDQMClientHeavyIons.cc  modified in PR(s): #28342
  - File  DQMServices/ClientConfig/interface/DQMGenericClient.h  modified in PR(s): #28247
  - File  DQMServices/Components/plugins/SealModule.cc  modified in PR(s): #28295
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/QReport.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/interface/QTest.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/src/DQMStore.cc  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #28092, #28247, #28342

cmsbuild commented:
Pull request #28316 was updated. @SiewYan, @andrius-k, @schneiml, @Martin-Grunewald, @rekovic, @fioriNTU, @tlampen, @alberto-sanchez, @pohsun, @santocch, @peruzzim, @cmsbuild, @agrohsje, @fwyzard, @efeyazgan, @tocheng, @jfernan2, @qliphy, @benkrikler, @mkirsano, @kmaeshima, @christopheralanwest, @franzoni, @fgolf can you please check and sign again.

schneiml committed:
Turn off end-job QTests by default.

Probably makes less sense but might give cleaner PR test results.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12624

- This PR adds an extra 388KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQM/TrackingMonitorClient/plugins/TrackingDQMClientHeavyIons.cc  modified in PR(s): #28342
  - File  DQMServices/ClientConfig/interface/DQMGenericClient.h  modified in PR(s): #28247
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/QReport.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/interface/QTest.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/src/DQMStore.cc  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #28092, #28247, #28342

cmsbuild commented:
Pull request #28316 was updated. @SiewYan, @andrius-k, @schneiml, @Martin-Grunewald, @rekovic, @fioriNTU, @tlampen, @alberto-sanchez, @pohsun, @santocch, @peruzzim, @cmsbuild, @agrohsje, @fwyzard, @efeyazgan, @tocheng, @jfernan2, @qliphy, @benkrikler, @mkirsano, @kmaeshima, @christopheralanwest, @franzoni, @fgolf can you please check and sign again.

schneiml commented:
please test

rebased again, the `SiPixelActionExecutor` seems to work again, and let's see if not running qtests on endJob by default is closer to the old behaviour.

cmsbuild commented:
The tests are being triggered in jenkins.

schneiml committed:
Implement SiStripConfigParser.

cmsbuild commented:
+1
Tested at: 95ad88546d385d114781ab5b1d3b886d62ffa4d1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f137dc/3348/summary.html

cmsbuild commented:
Comparison job queued.

schneiml committed:
FIx a bug in QTest config parsing.

`set*Range` actually has the side-effect of enabling the check, so we
must not call it when the range indicates the check should not be used.

schneiml commented:
The latest update brings an (untested, not sure which sequence would use it) config parser  for `SiStripActionExecutor` and a bugfix that should resolve another pile of comparison failures.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f137dc/3348/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939026
 - DQMHistoTests: Total failures: 40
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2938645
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1200.517 KiB( 33 files compared)
 - DQMHistoSizes: changed ( 1000.0,... ): -0.588 KiB       Muons/Tests
 - DQMHistoSizes: changed ( 1000.0,... ): 0.533 KiB        Tracking/TrackParameters
 - DQMHistoSizes: changed ( 1000.0,... ): -0.093 KiB       L1T/L1TOccupancy
 - DQMHistoSizes: changed ( 10024.0,... ): -14.303 KiB      HLT/Higgs
 - DQMHistoSizes: changed ( 11634.0,... ): -35.896 KiB      HLT/Higgs
 - DQMHistoSizes: changed ( 136.731 ): 0.063 KiB        SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 136.788 ): 0.250 KiB        SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 136.85 ): 1.112 KiB        SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 140.56,... ): 0.668 KiB        Tracking/TrackParameters
 - DQMHistoSizes: changed ( 140.56 ): 0.059 KiB        SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 4.22 ):  ... <truncated>
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12631

- This PR adds an extra 392KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQM/TrackingMonitorClient/plugins/TrackingDQMClientHeavyIons.cc  modified in PR(s): #28342
  - File  DQMServices/ClientConfig/interface/DQMGenericClient.h  modified in PR(s): #28247
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/QReport.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/interface/QTest.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/src/DQMStore.cc  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #28092, #28247, #28342

cmsbuild commented:
Pull request #28316 was updated. @SiewYan, @andrius-k, @schneiml, @Martin-Grunewald, @rekovic, @fioriNTU, @tlampen, @alberto-sanchez, @pohsun, @santocch, @peruzzim, @cmsbuild, @agrohsje, @fwyzard, @efeyazgan, @tocheng, @jfernan2, @qliphy, @benkrikler, @mkirsano, @kmaeshima, @christopheralanwest, @franzoni, @fgolf can you please check and sign again.

schneiml committed:
EXPERIMENT: move to Leagcy module type

After the QTest changes the QualityTester runs too early in some
workflows, before other harvesting modules. Try to move back to a
edm::EDAnalyzer legacy base to see if that affects ordering.

Spoiler: seems it does not.

schneiml committed:
Add an ordering Dependency to QTests.

Now they reliably run after harvesting modules.

But somehow that does not fix the problem with HLT/Higgs/*Eff*?

schneiml committed:
Move DQMGenericClient to endRun.

This is a bit risky, but it fixes the order of operations for some cases
(HLT/Higgs efficiency QTests).

cmsbuild commented:
The code-checks are being triggered in jenkins.

schneiml commented:
Ok, after some long debugging, things got a bit more clear: Since the new QTests are only applied at discrete points in time (rather than "as needed", as the old implementation somehow did), we get into trouble with module ordering. Issue #28354 should explain the problem.

To work around that, I now moved the computations of `DQMGenericClient` to `endRun`, *and* enforce correct order in `endRun` using `DQMToken`s. No idea how much fallout that causes. 

Not let's do a round of tests and see how many differences remain (or which new ones appear).

Also, `SiStripConfigParser` is still untested, it seems to not run in any offline sequences.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12644

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQM/TrackingMonitorClient/plugins/TrackingDQMClientHeavyIons.cc  modified in PR(s): #28342
  - File  DQMServices/ClientConfig/interface/DQMGenericClient.h  modified in PR(s): #28247
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/QReport.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/interface/QTest.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/src/DQMStore.cc  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #28092, #28247, #28342


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12644/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12644/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12644/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12644/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

schneiml commented:
please test

let's see what happens now.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12646

- This PR adds an extra 400KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQM/TrackingMonitorClient/plugins/TrackingDQMClientHeavyIons.cc  modified in PR(s): #28342
  - File  DQMServices/ClientConfig/interface/DQMGenericClient.h  modified in PR(s): #28247
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/QReport.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/interface/QTest.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/src/DQMStore.cc  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #28092, #28247, #28342


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12646/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12646/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

schneiml commented:
please test

really, it should pass code-checks now.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12647

- This PR adds an extra 400KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQM/TrackingMonitorClient/plugins/TrackingDQMClientHeavyIons.cc  modified in PR(s): #28342
  - File  DQMServices/ClientConfig/interface/DQMGenericClient.h  modified in PR(s): #28247
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/QReport.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/interface/QTest.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/src/DQMStore.cc  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #28092, #28247, #28342

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #28316 was updated. @SiewYan, @andrius-k, @schneiml, @Martin-Grunewald, @rekovic, @fioriNTU, @tlampen, @alberto-sanchez, @pohsun, @santocch, @peruzzim, @cmsbuild, @agrohsje, @fwyzard, @efeyazgan, @tocheng, @jfernan2, @qliphy, @benkrikler, @mkirsano, @kmaeshima, @christopheralanwest, @franzoni, @fgolf can you please check and sign again.

cmsbuild commented:
+1
Tested at: 8c808d40dc7a0a2edf3c05263edb0e8cf3de3fcd
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f137dc/3370/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f137dc/3370/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939026
 - DQMHistoTests: Total failures: 73
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2938612
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -1101.681 KiB( 33 files compared)
 - DQMHistoSizes: changed ( 1000.0 ): -2.081 KiB       SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 1000.0,... ): -0.588 KiB       Muons/Tests
 - DQMHistoSizes: changed ( 1000.0,... ): 0.533 KiB        Tracking/TrackParameters
 - DQMHistoSizes: changed ( 1000.0,... ): -0.093 KiB       L1T/L1TOccupancy
 - DQMHistoSizes: changed ( 10224.0 ): -12.221 KiB      SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 136.731 ): -6.042 KiB       SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 136.788 ): -7.950 KiB       SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 136.85 ): -6.368 KiB       SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 140.56 ): -67.795 KiB      SiStrip/MechanicalView
 - DQMHistoSizes: changed ( 140.56,... ): 0.668 KiB        Tracking/TrackParameters
 - DQMHistoSizes: changed ( 250202.181 ):  ... <truncated>
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

schneiml commented:
The bad news: the `SiStripMonitorClient` bad modules list mechanism is broken.

The good news: it looks like *only* the bad modules list mechanism is broken.

Actually, I am mildly puzzled. The mechanism in general was working well before, only a few QTests misfiring due to mir-read configuration caused too many bad modules. Now, it seems we have too little bad modules, but the thing that really changed was the `DQMGenericClient`? or maybe the ordering dependency. I'll investigate.

schneiml committed:
Move muon efficiencies to endRun.

This should fix the remaining, missing QTests. By creating the MEs in
endRun, they can be QTest'ed in endRun (with dependencies) and then the
QReports read in endJob (though actually, the code in question seems to
not do much with the QReports).

schneiml committed:
Make SiStripOfflineDQM explicitly depend on its QualityTester.

This should make sure that the QReports are available by the time that
SiStripOfflineDQM analyzes them in endRun.

Except, it seems to not work, and the modules still run in reverse
order.

The change to edm::one is to take the more complicated edm legacy
semantics out of the equations, but does not seem to change anything.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12655

- This PR adds an extra 416KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQM/TrackingMonitorClient/BuildFile.xml  modified in PR(s): #28357
  - File  DQM/TrackingMonitorClient/plugins/BuildFile.xml  modified in PR(s): #28357
  - File  DQM/TrackingMonitorClient/plugins/TrackingDQMClientHeavyIons.cc  modified in PR(s): #28342
  - File  DQMServices/ClientConfig/interface/DQMGenericClient.h  modified in PR(s): #28247
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/QReport.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/interface/QTest.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/src/DQMStore.cc  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #28092, #28247, #28342

cmsbuild commented:
Pull request #28316 was updated. @SiewYan, @andrius-k, @schneiml, @Martin-Grunewald, @rekovic, @fioriNTU, @tlampen, @alberto-sanchez, @pohsun, @santocch, @peruzzim, @cmsbuild, @agrohsje, @fwyzard, @efeyazgan, @tocheng, @jfernan2, @qliphy, @benkrikler, @mkirsano, @kmaeshima, @christopheralanwest, @franzoni, @fgolf can you please check and sign again.

schneiml commented:
please test

Now, commit a9a0dbc should fix the `-8` in many different workflows, which are related to some muon plots booked in harvesting. Commit 29c77f3 should fix the 100's-1000's of missing SiStrip BadModuleList entries in some workflows, except apparently it does not, due to unexpected behaviour of EDM reported in issue #28354. Let's see what happens.

cmsbuild commented:
The tests are being triggered in jenkins.

schneiml commented:
please abort

found a way that actually works, coming soon

cmsbuild commented:
Jenkins tests are aborted.

schneiml committed:
Depend on *all* DQMTokens.

This is stupid and wrong, but actually works.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28316/12657

- This PR adds an extra 416KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQM/TrackingMonitorClient/BuildFile.xml  modified in PR(s): #28357
  - File  DQM/TrackingMonitorClient/plugins/BuildFile.xml  modified in PR(s): #28357
  - File  DQM/TrackingMonitorClient/plugins/TrackingDQMClientHeavyIons.cc  modified in PR(s): #28342
  - File  DQMServices/ClientConfig/interface/DQMGenericClient.h  modified in PR(s): #28247
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/interface/QReport.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/interface/QTest.h  modified in PR(s): #28092, #28342
  - File  DQMServices/Core/src/DQMStore.cc  modified in PR(s): #28092, #28247, #28342
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #28092, #28247, #28342

cmsbuild commented:
Pull request #28316 was updated. @SiewYan, @andrius-k, @schneiml, @Martin-Grunewald, @rekovic, @fioriNTU, @tlampen, @alberto-sanchez, @pohsun, @santocch, @peruzzim, @cmsbuild, @agrohsje, @fwyzard, @efeyazgan, @tocheng, @jfernan2, @qliphy, @benkrikler, @mkirsano, @kmaeshima, @christopheralanwest, @franzoni, @fgolf can you please check and sign again.

schneiml commented:
please test

now we got chances for a pretty clean result.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7add316de292eb7bedb44b39d1bfc2a99688803b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f137dc/3379/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f137dc/3379/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2936360
 - DQMHistoTests: Total failures: 41
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2935978
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 12.064 KiB( 33 files compared)
 - DQMHistoSizes: changed ( 1000.0,... ): 0.533 KiB        Tracking/TrackParameters
 - DQMHistoSizes: changed ( 1000.0,... ): -0.093 KiB       L1T/L1TOccupancy
 - DQMHistoSizes: changed ( 140.56,... ): 0.668 KiB        Tracking/TrackParameters
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

schneiml commented:
And indeed, things look much better. The *changes* that we now see need further investigation. The additional elements are as far as I remember QTests in Tracking which seem to be quite normal, not sure why they did not run in the past, and I don't think I care too much.

schneiml commented:
Ok, `SiStripCertificationInfo` is pretty much unfixable until we learn more about #28354. I tried moving it back by `consumesMany<DQMToken, edm::InRun>` like I did for `siStripOfflineAnalyser`, but it gets scheduled right *before* `siStripOfflineAnalyser`, where it probably should go *after* it to work correctly. I am not sure if there currently *is* a way to schedule things manually, and if there is, I have no idea how to do it.

Note that implementing #28364 will *not* fix this problem, here we need manual scheduling of legacy modules. For all I know, this is not a new problem; it just happened to work *by chance* in the past. Even when I go back to legacy `EDAnalyzer` for the two modules in question, EDM now orders them exactly the wrong way:
```
++++++ starting: global end run for module: label = 'siStripQTester' id = 9
++++++ finished: global end run for module: label = 'siStripQTester' id = 9
++++++ starting: global end run for module: label = 'siStripCertificationInfo' id = 65
++++++ finished: global end run for module: label = 'siStripCertificationInfo' id = 65
++++++ starting: global end run for module: label = 'siStripOfflineAnalyser' id = 10
++++++ finished: global end run for module: label = 'siStripOfflineAnalyser' id = 10
```

Before, they where just ordered by module ID (it seems), which would be perfectly correct.

schneiml commented:
superseded by #28379 (+ #28375, + another one not opened yet)
